File: backend/app/schemas/employee.py
from datetime import date, datetime
from typing import Optional

from pydantic import BaseModel, EmailStr, Field


class EmployeeCreate(BaseModel):
    # tenant_id is NOT accepted from the client — it comes from the JWT.
    # Structural / locked fields stay required at the schema level.
    payroll_id: str = Field(min_length=1)
    employee_name: str = Field(min_length=1)
    date_of_birth: date  # JSON sends "YYYY-MM-DD"; Pydantic parses it.
    primary_store_id: int
    position_id: int
    # The brand the staffer belongs to (must be one the primary store serves —
    # its own brand, or a foodmall extra brand). Optional; defaults to the store's
    # primary brand when omitted.
    brand_id: Optional[int] = None
    # Configurable fields — required-ness is enforced per the form config,
    # so they are optional at the schema level.
    email: Optional[EmailStr] = None
    # When true and no email is given, the record is created "email pending"
    # (allowed only if the email_unavailable_enabled setting is on).
    email_pending: bool = False
    payrate: Optional[float] = Field(default=None, ge=0)
    pay_currency: Optional[str] = None  # e.g. "TTD", "USD", "JAM", "XCD"
    phone_number: Optional[str] = None
    mag_code: Optional[str] = None
    country_id: Optional[int] = None
    additional_store_ids: Optional[list[int]] = None


class ReviewUpdate(BaseModel):
    reviewed: bool


class ReviewFlagUpdate(BaseModel):
    # Field keys to flag as needing review. Empty list clears the flag.
    fields: list[str] = []


class MagUpdate(BaseModel):
    mag_code: Optional[str] = None


class EmployeeRead(BaseModel):
    employee_id: int
    tenant_id: int
    payroll_id: str
    employee_name: str
    date_of_birth: date
    phone_number: Optional[str]
    email: Optional[str]  # plain str on read, not EmailStr
    email_pending: bool = False
    payrate: Optional[float]
    pay_currency: Optional[str]
    mag_code: Optional[str]
    country_id: Optional[int]
    primary_store_id: Optional[int]
    position_id: Optional[int]
    brand_id: Optional[int]  # effective brand (own, else primary store's)
    reviewed: bool
    promotion_pending_review: bool = False
    review_fields: list[str] = []  # cells an Admin/IT flagged for review
    created_at: datetime
    created_by: Optional[int] = None
    reviewed_at: Optional[datetime] = None

    # Resolved display names (option b) — populated by the endpoint.
    store_name: Optional[str] = None
    brand_name: Optional[str] = None
    position_title: Optional[str] = None
    country_name: Optional[str] = None
    created_by_name: Optional[str] = None  # username of the creator
    additional_stores: list[str] = []  # resolved additional-store names
    additional_store_ids: list[int] = []  # ids, for editing

File: frontend/lib/models/employee.dart
class Employee {
  final int employeeId;
  final String payrollId;
  final String employeeName;
  final DateTime dateOfBirth;
  final String? phoneNumber;
  final String? email;
  final double? payrate;
  final String? payCurrency;
  final String? magCode;
  final int? primaryStoreId;
  final int? positionId;
  final int? countryId;
  final bool reviewed;
  final String? storeName;
  final String? brandName;
  final String? positionTitle;
  final String? countryName;
  final List<String> additionalStores;
  final List<int> additionalStoreIds;

  const Employee({
    required this.employeeId,
    required this.payrollId,
    required this.employeeName,
    required this.dateOfBirth,
    required this.phoneNumber,
    required this.email,
    required this.payrate,
    required this.payCurrency,
    required this.magCode,
    required this.primaryStoreId,
    required this.positionId,
    required this.countryId,
    required this.reviewed,
    required this.storeName,
    required this.brandName,
    required this.positionTitle,
    required this.countryName,
    this.additionalStores = const [],
    this.additionalStoreIds = const [],
  });

  factory Employee.fromJson(Map<String, dynamic> j) => Employee(
        employeeId: j['employee_id'] as int,
        payrollId: j['payroll_id'] as String,
        employeeName: j['employee_name'] as String,
        dateOfBirth: DateTime.parse(j['date_of_birth'] as String),
        phoneNumber: j['phone_number'] as String?,
        email: j['email'] as String?,
        payrate: (j['payrate'] as num?)?.toDouble(),
        payCurrency: j['pay_currency'] as String?,
        magCode: j['mag_code'] as String?,
        primaryStoreId: j['primary_store_id'] as int?,
        positionId: j['position_id'] as int?,
        countryId: j['country_id'] as int?,
        reviewed: (j['reviewed'] as bool?) ?? false,
        storeName: j['store_name'] as String?,
        brandName: j['brand_name'] as String?,
        positionTitle: j['position_title'] as String?,
        countryName: j['country_name'] as String?,
        additionalStores: (j['additional_stores'] as List?)
                ?.map((e) => e.toString())
                .toList() ??
            const [],
        additionalStoreIds: (j['additional_store_ids'] as List?)
                ?.map((e) => e as int)
                .toList() ??
            const [],
      );

  String get additionalStoresDisplay =>
      additionalStores.isEmpty ? '—' : additionalStores.join(', ');

  /// DOB shown US-style MM/DD/YYYY across the app.
  String get dobDisplay =>
      '${dateOfBirth.month.toString().padLeft(2, '0')}/'
      '${dateOfBirth.day.toString().padLeft(2, '0')}/'
      '${dateOfBirth.year}';

  String get payrateDisplay {
    if (payrate == null) return '—';
    final amount = payrate!.toStringAsFixed(2);
    return (payCurrency == null || payCurrency!.isEmpty)
        ? amount
        : '$amount $payCurrency';
  }
}
